### DISCUSSION TITLE
Add country to production and edit it for localhost/CI

### DISCUSSION THREAD

**Discussion by sspooky13** [shopsys/shopsys#2242] [Q&A]
Hello, 
In one unnamed project, I need to add country to database. So, I created data in migration for tables `countries`, `country_domains`, `country_translations`. This is ok.
Issue is in localhost/CI where after running `db-demo` I have only first domain in migrations because `baseUrl` in `setting_values` exists only for first domain. After running migrations is running `php phing domains-data-create` where is added `null` to name, so smoke tests throw fatal `TypeError: Return value of Shopsys\FrameworkBundle\Model\Country\CountryTranslation::getName() must be of the type string, null returned` on [this line](https://github.com/shopsys/shopsys/blob/master/packages/framework/src/Model/Country/CountryTranslation.php#L32). So, I wanted to edit this country in CountryDataFixture class but there is another fatal `[Shopsys\FrameworkBundle\Model\Country\Exception\CountryDomainNotFoundException]  
  CountryDomain for country with ID 1 and domain ID 2 not found. ` because `CountryDomain` is created **only** in construct, not in edit (code try to edit CountryDomain[secondDomainId] and that instance not exists).

I solved this issue with native query in data fixtures but I want your opinion on what do you think is right solution? 🙏 

Thanks for your time.

PS: I talk about SSFW 8 project, maybe in SSFW 9 this issue not exists (but I don't think it's solved in SSFW 9).

### To Reproduce
1. Add any country in migrations
2. See error in smoke test
```
1) Tests\ShopBundle\Smoke\Http\HttpSmokeTest::testHttpResponse with data set #250 (Shopsys\HttpSmokeTesting\RequestDataSet Object (...))
TypeError: Return value of Shopsys\FrameworkBundle\Model\Country\CountryTranslation::getName() must be of the type string, null returned
/var/www/html/vendor/shopsys/framework/src/Model/Country/CountryTranslation.php:36
/var/www/html/vendor/shopsys/framework/src/Model/Country/Country.php:78
.
.
.
2) Tests\ShopBundle\Smoke\Http\HttpSmokeTest::testHttpResponse with data set #251 (Shopsys\HttpSmokeTesting\RequestDataSet Object (...))
TypeError: Return value of Shopsys\FrameworkBundle\Model\Country\CountryTranslation::getName() must be of the type string, null returned
/var/www/html/vendor/shopsys/framework/src/Model/Country/CountryTranslation.php:36
/var/www/html/vendor/shopsys/framework/src/Model/Country/Country.php:78
.
.
.
--
There was 1 failure:
1) Tests\ShopBundle\Smoke\Http\HttpSmokeTest::testHttpResponse with data set #154 (Shopsys\HttpSmokeTesting\RequestDataSet Object (...))
Failed asserting that status code 500 for route "admin_country_edit" is identical to expected 200
Notes for this data set:
  - Route requires ID parameter "id". Using 1 by default.
  - Log as "admin" to administration.
Failed asserting that 500 is identical to 200.
/var/www/html/vendor/shopsys/http-smoke-testing/src/HttpSmokeTestCase.php:149
/var/www/html/vendor/shopsys/http-smoke-testing/src/HttpSmokeTestCase.php:52
```

1. Add any country in migrations
2. Edit country in `CountryDataFixture`
3. See error in `php phing db-fixtures-demo`
```
   > loading Shopsys\ShopBundle\DataFixtures\Demo\AdministratorDataFixture
   > loading Shopsys\ShopBundle\DataFixtures\Demo\CountryDataFixture
07:37:04 ERROR     [app] Shopsys\FrameworkBundle\Model\Country\Exception\CountryDomainNotFoundException: CountryDomain for country with ID 1 and domain ID 2 not found. (uncaught exception) at /var/www/html/vendor/shopsys/framework/src/Model/Country/Country.php line 172 while running console command `doctrine:fixtures:load` ["exception" => Shopsys\FrameworkBundle\Model\Country\Exception\CountryDomainNotFoundException { …}]
07:37:04 ERROR     [console] Error thrown while running command "doctrine:fixtures:load --append --no-interaction". Message: "CountryDomain for country with ID 1 and domain ID 2 not found." ["exception" => Shopsys\FrameworkBundle\Model\Country\Exception\CountryDomainNotFoundException { …},"command" => "doctrine:fixtures:load --append --no-interaction","message" => "CountryDomain for country with ID 1 and domain ID 2 not found."]
07:37:04 ERROR     [app] Command `doctrine:fixtures:load` exited with status code 1
In Country.php line 172:
                                                                                    
  [Shopsys\FrameworkBundle\Model\Country\Exception\CountryDomainNotFoundException]  
  CountryDomain for country with ID 1 and domain ID 2 not found.                    
                                                                                    
Exception trace:
  at /var/www/html/vendor/shopsys/framework/src/Model/Country/Country.php:172
 Shopsys\FrameworkBundle\Model\Country\Country->getCountryDomain() at /var/www/html/vendor/shopsys/framework/src/Model/Country/Country.php:87
 Shopsys\FrameworkBundle\Model\Country\Country->isEnabled() at /var/www/html/vendor/shopsys/framework/src/Model/Country/CountryDataFactory.php:60
 Shopsys\FrameworkBundle\Model\Country\CountryDataFactory->fillFromCountry() at /var/www/html/src/Shopsys/ShopBundle/Model/Country/CountryDataFactory.php:45
 Shopsys\ShopBundle\Model\Country\CountryDataFactory->fillFromCountry() at /var/www/html/src/Shopsys/ShopBundle/Model/Country/CountryDataFactory.php:34
 Shopsys\ShopBundle\Model\Country\CountryDataFactory->createFromCountry() at /var/www/html/src/Shopsys/ShopBundle/DataFixtures/Demo/CountryDataFixture.php:120
 Shopsys\ShopBundle\DataFixtures\Demo\CountryDataFixture->updateNetherlandsCountry() at /var/www/html/src/Shopsys/ShopBundle/DataFixtures/Demo/CountryDataFixture.php:103
 Shopsys\ShopBundle\DataFixtures\Demo\CountryDataFixture->load() at /var/www/html/vendor/doctrine/data-fixtures/lib/Doctrine/Common/DataFixtures/Executor/AbstractExecutor.php:114
 Doctrine\Common\DataFixtures\Executor\AbstractExecutor->load() at /var/www/html/vendor/doctrine/data-fixtures/lib/Doctrine/Common/DataFixtures/Executor/ORMExecutor.php:69
 Doctrine\Common\DataFixtures\Executor\ORMExecutor::Doctrine\Common\DataFixtures\Executor\{closure}() at n/a:n/a
 call_user_func() at /var/www/html/vendor/shopsys/doctrine-orm/lib/Doctrine/ORM/EntityManager.php:235
 Doctrine\ORM\EntityManager->transactional() at /var/www/html/vendor/shopsys/doctrine-orm/lib/Doctrine/ORM/Decorator/EntityManagerDecorator.php:76
 Doctrine\ORM\Decorator\EntityManagerDecorator->transactional() at /var/www/html/vendor/doctrine/data-fixtures/lib/Doctrine/Common/DataFixtures/Executor/ORMExecutor.php:71
 Doctrine\Common\DataFixtures\Executor\ORMExecutor->execute() at /var/www/html/vendor/doctrine/doctrine-fixtures-bundle/Command/LoadDataFixturesDoctrineCommand.php:127
 Doctrine\Bundle\FixturesBundle\Command\LoadDataFixturesDoctrineCommand->execute() at /var/www/html/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /var/www/html/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:1001
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:86
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:255
 Symfony\Component\Console\Application->doRun() at /var/www/html/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:74
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/html/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:148
 Symfony\Component\Console\Application->run() at /var/www/html/app/Bootstrap.php:64
 Shopsys\Bootstrap->run() at /var/www/html/bin/console:17
doctrine:fixtures:load [--append] [--group GROUP] [--em EM] [--shard SHARD] [--purge-with-truncate] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command>
```

**Comment by sspooky13** [comment]
Next issue that I have is run migration which creates data for new created domain. I need locale that new domain so I called `get_domain_locale($domainId)` and $domainId is from `getAllDomainIds()` (both methods is from `MultidomainMigrationTrait`).
But `get_domain_locale()` is returned with ERROR because in that function doesn't exist 4. domain and that function will be upgraded after migrations (it's phing command `domains-db-functions-create` which is running in command `domains-data-create` and all this is called after [migrations](https://github.com/shopsys/shopsys/blob/master/packages/framework/build.xml#L131))

Any idea of right way to prevent this issue? 🙏 
```
[PDOException (P0001)]                                                       
  SQLSTATE[P0001]: Raise exception: 7 ERROR:  Domain with ID 4 does not exist  
  s                                                                            
  CONTEXT:  PL/pgSQL function get_domain_locale(integer) line 7 at RAISE
```

**Comment by s3tezsky** [comment]
Hi @sspooky13,

I have to admit it is really hard to help you here. I think you are describing 3 different problems...

The first one:

I have added migrations which is creating
- new row in `countries`
- new rows in `country_domain` (for each domain)
- new rows in `country_translation` (for each locale)
By doing so, everything works without any issue (tested on `9.1`)

The second and third one:

It seems you should either do everything in migration and do not to let everything for `./phing domains-data-create`. Or do it in more iterative way (do not try to do all the stuff at once - but split it into smaller steps - create new domain, run domains-data-create, add country by migration, and so on).

I will not probably not anwer to all your question but hope it will help you too.

**Reply by sspooky13** [reply]
Hello @s3tezsky,
I forgot to say, I wanted to create new domain, so try to add new domain and now run that migration.

I think we don't understand what is real scenario. I will try to describe it.
Imagine you need to add new domain and every domain has default country and that country for new domain doesn't exist so you want to create it. Everything seems ok but issues are what I describe above ⬆.

Splitting this into smaller steps doesn't help because you do deploy all in once. Only this can help if you deploy new domain gradually by steps.

This is quite hard to describe, so I suggest calling about this and I can show you this issue in project.

**Reply by s3tezsky** [reply]
I dig a bit around and find out it is bug which was fixed in #2191. It will keeps your Country entity defined in a good states so your smoke test should pass then.

I am still looking into `get_domain_locale` function in postgres... I will update this thread when I will have anything usefull.

**Reply by s3tezsky** [reply]
UPDATE: For more reliable migrations when adding a new domain we have decided to deprecate `MultidomainMigrationTrait` and rework behavior to get domain informations right from `domains.yaml` instead of database functions. See #2270 for more informations.
